tooling/helpers/range.go

```diff
@@ -0,0 +1,247 @@
+				),
+				test.Expect().Status(http.StatusPartialContent).Body(
+					check.And(
+						append([]check.Check[string]{check.Contains("Content-Type: application/vnd.ipld.raw")}, multirangeBodyChecks...)...,
+					),
+				).Headers(test.Header("Content-Type").Contains("multipart/byteranges")),
+			),
+		),
+	}
+
+	return rangeTest
+}
+
+// BaseWithRangeTestTransform takes a test where there is no "Range" header set in the request, or checks on the
+// StatusCode, Body, or Content-Range headers and verifies whether a valid response is given for the requested ranges.
+// Will test the full request, a single range request for the first passed range as well as a multi-range request for
+// all the requested ranges.
+//
+// If no ranges are passed then some non-overlapping ranges are automatically generated for data >= 10 bytes. Smaller
+// data will result in undefined behavior.
+//
+// Note: HTTP Range requests can be validly responded with either the full data, or the requested partial data
+// Note: HTTP Multi Range requests can be validly responded with one of the full data, the partial data from the first
+// range, or the partial data from all the requested ranges
+func BaseWithRangeTestTransform(t *testing.T, baseTest test.SugarTest, branges ByteRanges, fullData []byte) test.SugarTests {
+	standardBase := test.SugarTest{
+		Name:     fmt.Sprintf("%s - full request", baseTest.Name),
+		Hint:     baseTest.Hint,
+		Request:  baseTest.Request,
+		Requests: baseTest.Requests,
+		Response: test.AllOf(
+			baseTest.Response,
+			test.Expect().Status(http.StatusOK).Body(fullData),
+		),
+		Responses: baseTest.Responses,
+	}
+	rangeTests := RangeTestTransform(t, baseTest, branges, fullData)
+	return append(test.SugarTests{standardBase}, rangeTests...)
+}
+
+// RangeTestTransform takes a test where there is no "Range" header set in the request, or checks on the
+// StatusCode, Body, or Content-Range headers and verifies whether a valid response is given for the requested ranges.
+// Will test both a single range request for the first passed range as well as a multi-range request for all the
+// requested ranges.
+//
+// If no ranges are passed then some non-overlapping ranges are automatically generated for data >= 10 bytes. Smaller
+// data will result in undefined behavior.
+//
+// Note: HTTP Range requests can be validly responded with either the full data, or the requested partial data
+// Note: HTTP Multi Range requests can be validly responded with one of the full data, the partial data from the first
+// range, or the partial data from all the requested ranges
+func RangeTestTransform(t *testing.T, baseTest test.SugarTest, branges ByteRanges, fullData []byte) test.SugarTests {
+	if len(branges) == 0 {
+		dataLen := len(fullData)
+		if dataLen < 10 {
+			panic("transformation not defined for data smaller than 10 bytes")
+		}
+
+		branges = ByteRanges{
+			SimpleByteRange(7, 9),
+			SimpleByteRange(1, 3),
+		}
+	}
```

Comment:
Having some flexibility here seemed useful, I was finding that coming up with these numbers when they were in the tests was mostly a repeated exercise around what numbers wouldn't panic and what might be some wacky orders to try.

We could make this randomized, use the data size to determine the ranges, try things like suffix requests, etc. This is mostly for illustration and was a trivial example. Suggestions welcome.

Reply:
Long answer on purpose, I'll probably reuse this later:

The idea behind this test suite is to have a test suite that resembles the specs, and that can be read, understood, and written by someone who is not a maintainer or even a go programmer. If you can read English, you can understand the specs and the test definitions.

When we move test-related data out of the test definition (or we add memory management tidbits see my comment below), we change the nature of the project.

To me, the architecture of the conformance test should be:

- ./tests => targets humans, maps to specs, self-contain, written with English-like sugar
- ./tooling/helpers => targets devs, data transformation operations, based on the lower-APIs, written with some go knowledge. Expose English-API to humans reading `./tests`
- ./tooling => targets go maintainers, yolo

That split is one of the reasons we didn't "just" translate shell into go code but went for a data-driven approach.

That code seems to move test data out of the test definition for the sake of making it generalizable and saving lines of go code. Could we define this as a const in the test file itself and reuse it?